test/decimal_test.exs

```diff
@@ -119,6 +119,20 @@ defmodule DecimalTest do
     assert Decimal.from_float(-1.5) == d(-1, 15, -1)
   end
 
+  test "cast/1" do
+    assert Decimal.cast(123) == {:ok, d(1, 123, 0)}
+    assert Decimal.cast(123.0) == {:ok, d(1, 1230, -1)}
+    assert Decimal.cast("123") == {:ok, d(1, 123, 0)}
+    assert Decimal.cast(d(1, 123, 0)) == {:ok, d(1, 123, 0)}
+
+    assert Decimal.cast("one two three") == :error
+    assert Decimal.cast("e0") == :error
+
+    assert_raise FunctionClauseError, fn ->
+      Decimal.cast(:one_two_three)
```

Comment:
we could also return `:error` here instead. I'm fine with either approach.

Reply:
If we do return `:error` we should change the spec to accept any input type.